File: Orimath.ViewPlugins/Plugins/IDispatcher.fs
namespace Orimath.Plugins
open System.Threading

type IDispatcherInvoker =
    abstract member Invoke : action: (unit -> unit) -> unit


type IDispatcher =
    abstract member SyncContext : SynchronizationContext
    abstract member Background : IDispatcherInvoker
    abstract member UI : IDispatcherInvoker


[<AutoOpen>]
module DispatcherOperator =
    type IDispatcherInvoker with
        member inline _.Bind(m: IDispatcherInvoker, f) = m.Invoke(f)
        member inline _.Zero() = ()
        member inline this.Delay(f: unit -> unit) = this.Invoke(f)

File: Orimath.ViewPlugins/Plugins/IDispatcher.cs
using System;
using System.Threading.Tasks;
using System.Threading;

namespace Orimath.Plugins
{
    public interface IDispatcher
    {
        SynchronizationContext SynchronizationContext { get; }
        void OnBackgroundAsync(Action action);
        void OnUIAsync(Action action);
    }
}
